core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/AsyncRetryableStage.java

```diff
@@ -112,6 +124,24 @@ private RetryExecutor(SdkHttpFullRequest request, RequestExecutionContext contex
 
         public CompletableFuture<Response<OutputT>> execute() throws Exception {
             CompletableFuture<Response<OutputT>> future = new CompletableFuture<>();
+
+            long apiCallTimeoutInMillis = getClientExecutionTimeoutInMillis(context.requestConfig());
+            if (apiCallTimeoutInMillis > 0) {
+                TimeoutTask timeout = new AsyncTimeoutTask(future, ApiCallTimeoutException.create());
+                ScheduledFuture<?> scheduledFuture =
+                    timeoutExecutor.schedule(timeout,
+                                             apiCallTimeoutInMillis,
+                                             TimeUnit.MILLISECONDS);
+                TimeoutTracker timeoutTracker = new ApiCallTimeoutTracker(timeout, scheduledFuture);
+                context.apiCallTimeoutTracker(timeoutTracker);
+
+                future.whenComplete((o, t) -> {
+                    if (t != null) {
```

Comment:
We need the checks here because we don't want it to timeout if the completableFuture fails exceptionally. I'll add a test for it.

Reply:
If the future has already been completed, the timeout task should not have any effect though right? Unless it tries to force the timeout exception by using `obtrudeException`: https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html#obtrudeException-java.lang.Throwable-

Reply:
Whoops.. I meant to write `if (t == null)`(The unit tests succeeded because there were either response unmarshaller expception or ApiCallAttemptTimeoutException got retried). But yeah, I think you are right, we should remove the check here.